File: multisnaprecyclerview/src/main/java/com/takusemba/multisnaprecyclerview/StartSnapHelperDelegator.java
package com.takusemba.multisnaprecyclerview;

import androidx.recyclerview.widget.OrientationHelper;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

/**
 * {@link SnapHelperDelegator} for start gravity
 **/
class StartSnapHelperDelegator extends SnapHelperDelegator {

  /**
   * Constructor
   *
   * @param snapCount the number of items to scroll over
   */
  StartSnapHelperDelegator(int snapCount) {
    super(snapCount);
  }

  @Override int getDistance(RecyclerView.LayoutManager layoutManager, View targetView,
      OrientationHelper helper) {
    final int childStart = getChildPosition(targetView, helper);
    final int containerStart = getContainerPosition(layoutManager, helper);
    return childStart - containerStart;
  }

  @Override int getContainerPosition(RecyclerView.LayoutManager layoutManager,
      OrientationHelper helper) {
    return helper.getStartAfterPadding();
  }

  @Override int getChildPosition(View targetView, OrientationHelper helper) {
    return helper.getDecoratedStart(targetView);
  }

  @Override boolean shouldSkipTarget(View targetView, RecyclerView.LayoutManager layoutManager,
      OrientationHelper helper, boolean forwardDirection) {
    return forwardDirection ? getDistance(layoutManager, targetView, helper) < 0
        : getDistance(layoutManager, targetView, helper) > 0;
  }
}

File: multisnaprecyclerview/src/main/java/com/takusemba/multisnaprecyclerview/StartSnapHelperDelegator.kt
package com.takusemba.multisnaprecyclerview

import android.view.View
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * [SnapHelperDelegator] for start gravity
 *
 * @param snapCount the number of items to scroll over
 */
class StartSnapHelperDelegator(snapCount: Int) : SnapHelperDelegator(snapCount) {

  override fun getDistance(
      layoutManager: RecyclerView.LayoutManager, targetView: View,
      helper: OrientationHelper
  ): Int {
    val childStart = getChildPosition(targetView, helper)
    val containerStart = getContainerPosition(layoutManager, helper)
    return childStart - containerStart
  }

  override fun getContainerPosition(
      layoutManager: RecyclerView.LayoutManager,
      helper: OrientationHelper
  ): Int {
    return helper.startAfterPadding
  }

  override fun getChildPosition(targetView: View, helper: OrientationHelper): Int {
    return helper.getDecoratedStart(targetView)
  }

  override fun shouldSkipTarget(
      targetView: View, layoutManager: RecyclerView.LayoutManager,
      helper: OrientationHelper, forwardDirection: Boolean
  ): Boolean {
    return if (forwardDirection)
      getDistance(layoutManager, targetView, helper) < 0
    else
      getDistance(layoutManager, targetView, helper) > 0
  }
}
